### DISCUSSION TITLE
Problems with youtube (Bookwidgets & SEB) - Some work - some don't

### DISCUSSION THREAD

**Discussion by LodeSFX** [SafeExamBrowser/seb-win-refactoring#813] [Q&A] [answered]
We have been running into the same problem several times when using Youtube in tests with Bookwidgets.

To show with a concrete example:
Our colleague gave a test via Bookwidgets this morning and it included two movie clips from Youtube. Both times Youtube opened correctly, but the student received a message saying that both movies could not be played in this browser. While in the Youtube window, however, you can click on other videos, which then sometimes play? Why is it that some videos play and some videos do not play?

The test contained the following videos:

https://www.youtube.com/watch?v=L3vm5QXnRDk&t=32s (Then again, one of this owner's other videos works just fine, for example)
https://www.youtube.com/watch?v=VGZc_ScUK30

Both videos are in the public domain.

Thank you for your suggestions, help, ...

**Comment by dbuechel** [comment]
This most likely relates to the fact that SEB 3.x does not support proprietary codecs. This issue has been discussed here already, see e.g. #102.

**Reply by LodeSFX** [reply]
I thank you for your response and indeed I had seen this topic appear a few times, but not all people are familiar with all that jargon, of course. So I would have liked a somewhat clearer answer. What exactly are codecs? And is there certain way to know which movies will work in SEB and which will not work in SEB? Things like that.

Thanks!

**Reply by dbuechel** [reply]
With respect to video codecs in general, please refer to https://en.wikipedia.org/wiki/Video_codec. And yes, there is and that aspect is already explained in the linked issue, see https://github.com/SafeExamBrowser/seb-win-refactoring/issues/102#issuecomment-875753362.

**Reply by dbuechel** [reply]
See also https://github.com/SafeExamBrowser/seb-win-refactoring/issues/475#issuecomment-1262072713 which is linked in the issue specified above as well.

**Comment by xavier-bw** [comment, marked as answer]
I investigated the YouTube communication of a few YouTube videos that didn't play in SEB and some that play in SEB and this confirmed Damian's explanation.  

When launching a video YouTube sends a list of available versions for the video (different resolutions / different codecs).  In cases where the video doesn't play these all use a proprietary codec (avc1). In cases the video plays, a version using a non-proprietary codec (vp9) was also listed.  

Both videos mentioned above now play in SEB, which seems to point to the fact that the list of versions can change.  For all videos I tried, all browsers I tried received the same list of versions.

All of this matches what's observed: some videos will play / some won't.  

Googling about how to force google to make a vp9 version available I found some 3 year old posts describing what YouTube is doing based upon experiments people did.  I'm only describing what I read, I did not verify this. 

If you upload a 1440p or 4K video to YouTube, they state a vp9 version will always get created.  If you upload a lower resolution video, e.g. standard HD (1080x1920) or lower it may get encoded in vp9 at some point. I read that this may be related to the number of plays and popularity of the channel. Based upon this observation, if you have a lower-resolution video you want to upload to YouTube and use in SEB, you should re-encode the video to a higher resolution before upload.

**Reply by dbuechel** [reply]
Excellent work and thanks a lot for the information! This will indeed be very useful for a lot of other users when dealing with online videos resp. YouTube in particular.

**Reply by xavier-bw** [reply]
I now also tested the upload of a 30 second 1440p mp4 video to YouTube and can confirm that YouTube encoded it as vp9 and it worked in SEB 3.6.0..
